src/daemon/rpmostreed-deployment-utils.c

```diff
@@ -369,63 +368,218 @@ rpmostreed_deployment_generate_variant (OstreeSysroot *sysroot,
+  gboolean is_new_checksum = FALSE;
+  if (!new_checksum)
+    new_checksum = g_strdup (current_checksum);
   else
-    {
-      g_autoptr(RpmOstreeOrigin) origin = NULL;
+    is_new_checksum = !g_str_equal (new_checksum, current_checksum);
 
-      origin = rpmostree_origin_parse_deployment (deployment, error);
-      if (!origin)
-        return NULL;
-      origin_refspec = g_strdup (rpmostree_origin_get_refspec (origin));
-    }
+  g_autoptr(GVariant) commit = NULL;
+  if (!ostree_repo_load_variant (repo, OSTREE_OBJECT_TYPE_COMMIT,
+                                 new_checksum, &commit, error))
+    return FALSE;
 
-  g_assert (origin_refspec);
+  GVariantDict dict;
+  g_variant_dict_init (&dict, NULL);
 
-  /* allow_noent=TRUE since the ref may have been deleted for a
-   * rebase.
-   */
-  if (!ostree_repo_resolve_rev (repo, origin_refspec,
-                                TRUE, &head, error))
-    return NULL;
+  /* first get all the traditional/backcompat stuff */
+  if (!add_all_commit_details_to_vardict (deployment, repo, refspec,
+                                          new_checksum, commit, &dict, error))
+    return FALSE;
 
-  if (head == NULL)
-    head = g_strdup (ostree_deployment_get_csum (deployment));
+  /* This may seem trivial, but it's important to keep the final variant as self-contained
+   * and "diff-based" as possible, since it'll be available as a D-Bus property. This makes
+   * it easier to consume for UIs like GNOME Software and Cockpit. */
+  g_variant_dict_insert (&dict, "is-new-checksum", "b", is_new_checksum);
 
-  if (!ostree_repo_load_variant (repo,
-                                 OSTREE_OBJECT_TYPE_COMMIT,
-                                 head,
-                                 &commit,
-                                 error))
-    return NULL;
+  if (is_new_checksum)
+    {
+      /* note we allow_noent here; we'll just skip over the rpm diff if there's no data */
+      g_autoptr(GVariant) ostree_rpm_diff = NULL;
+      if (!rpm_ostree_db_diff_variant (repo, current_checksum, new_checksum, TRUE,
+                                       &ostree_rpm_diff, NULL, error))
+        return FALSE;
+
+      if (ostree_rpm_diff)
+        g_variant_dict_insert (&dict, "ostree-rpm-diff", "@a(sua{sv})",
+                               g_steal_pointer (&ostree_rpm_diff));
+    }
 
-  sigs = rpmostreed_deployment_gpg_results (repo, origin_refspec, head, &gpg_enabled);
+  /* now we look at the rpm-md side */
 
-  g_variant_dict_init (&dict, NULL);
-  if (osname != NULL)
-    g_variant_dict_insert (&dict, "osname", "s", osname);
-  g_variant_dict_insert (&dict, "checksum", "s", head);
-  variant_add_commit_details (&dict, NULL, commit);
-  g_variant_dict_insert (&dict, "origin", "s", origin_refspec);
-  if (sigs != NULL)
-    g_variant_dict_insert_value (&dict, "signatures", sigs);
-  g_variant_dict_insert (&dict, "gpg-enabled", "b", gpg_enabled);
-  return g_variant_dict_end (&dict);
+#if 0
```

Comment:
I don't understand this part really at all.  Is it just to get a diff?

Reply:
Right, this is basically still incomplete, but it's where we'll get the layering level diff.